middle_end/flambda/terms/code_size.rec.ml

```diff
@@ -0,0 +1,503 @@
+  | Bigarray_set (_dims, (Complex32 | Complex64), _layout) ->
+    5 (* ~ 3 block_load + 2 block_set *)
+  | Bigarray_set (_dims, _kind, _layout) ->
+    2 (* ~ 1 block_load + 1 block_set *)
+
+let variadic_prim_size prim args =
+  match (prim : Flambda_primitive.variadic_primitive) with
+  | Make_block (_, _mut)
+  (* CR mshinwell: I think Make_array for a generic array ("Anything") is
+     more expensive than the other cases *)
+  | Make_array (_, _mut) -> alloc_size + List.length args
+
+let prim_size (prim : Flambda_primitive.t) =
+  match prim with
+  | Unary (p, _) -> unary_prim_size p
+  | Binary (p, _, _) -> binary_prim_size p
+  | Ternary (p, _, _, _) -> ternary_prim_size p
+  | Variadic (p, args) -> variadic_prim_size p args
+
+let simple simple = Simple.pattern_match simple ~const:(fun _ -> 1) ~name:(fun _ -> 0)
+
+let prim = prim_size
+
+let static_consts _ = 0
+
+(* The size of a set of closures is roughly the sum of the size of the
+   functions it refers to. This is mainly due to the fact that if we do inline
+   a function f where a set of closure is defined then we will copy the body of
+   all functions referred by this set of closure as they are dependent upon f.*)
+let set_of_closures ~find_code_size set_of_closures =
+  let func_decls = Set_of_closures.function_decls set_of_closures in
+  let funs = Function_declarations.funs func_decls in
+  Closure_id.Map.fold (fun _ func_decl size ->
+    let code_id = Function_declaration.code_id func_decl in
+    match find_code_size code_id with
+    | Or_unknown.Known s -> size + s
+    | Or_unknown.Unknown ->
+      Misc.fatal_errorf "Code size should have been computed for:@ %a"
+        Code_id.print code_id
+  )
+    funs (zero)
+
+let increase_due_to_let_expr ~size_of_defining_expr =
+  size_of_defining_expr
+
+let apply apply = 
+  match Apply.call_kind apply with
+  | Function Direct _ -> direct_call_size
+  (* CR mshinwell: Check / fix these numbers *)
+  | Function Indirect_unknown_arity -> indirect_call_size
+  | Function Indirect_known_arity _ -> indirect_call_size
+  | C_call { alloc = true; _ } -> alloc_extcall_size
+  | C_call { alloc = false; _ } -> nonalloc_extcall_size
+  | Method _ -> 8 (* from flambda/inlining_cost.ml *)
+
+let apply_cont apply_cont =
+  let size = match Apply_cont.trap_action apply_cont with
+    | None -> 0
+    | Some (Push _ | Pop _) -> 0 + 4
+  in
+  size + 1
+
+let invalid _ = 0
+
+let switch switch = 0 + (5 * Switch.num_arms switch)
+
+let increase_due_to_let_cont_non_recursive ~size_of_handler =
+  size_of_handler
+
+let increase_due_to_let_cont_recursive ~size_of_handlers =
+  size_of_handlers
+
+let rec expr_size ~find_code expr size =
+  match Expr.descr expr with
+  | Let let_expr ->
+    let size = named_size ~find_code (Let_expr.defining_expr let_expr) size in
+    Let_expr.pattern_match let_expr
+      ~f:(fun _bindable_let_bound ~body -> expr_size ~find_code body size)
```

Comment:
Phantom let-bindings should not be counted towards the size of the term. It's not strictly speaking related to this PR, but let's fix that while we're at it.